### PR TITLE
Add container environment variables correctly to the health check

### LINKED ISSUE #33021: HEALTHCHECK fails due to "invalid environment"
**Description**

In #31491 a change was made to bring environment variables into a HEALTHCHECK. However, proper checking is not done and if you have an environment variable without a value you get the following error:

```
rpc error: code = 2 desc = oci runtime error: exec failed: container_linux.go:247: starting container process caused \"invalid environment 'FOOBAR'\"\n"
```

In the file daemon/container.go: (https://github.com/moby/moby/blob/17.04.x/daemon/container.go#L240) there is a call to opts.ValidateEnv() call to ensure that no bad environment variables are passed down to the container. This should be done in daemon/health.go (https://github.com/moby/moby/blob/17.04.x/daemon/health.go#L81) in order to prevent health checks from failing when there is no value for an environment variable.

This is a problem from 17.03.1 onward.

This is preventing us from upgrading to the latest version of docker-ce, and a quick fix would be appreciated!

**Steps to reproduce the issue:**
1. Create an image with this Dockerfile
```
FROM alpine:latest
HEALTHCHECK --interval=5s --timeout=5s --retries=5 CMD /bin/sh -c "sleep 1"
ENTRYPOINT /bin/sh -c "sleep 600"
```
And build it
`docker build -t env_test .`

2. Run the container with the -e parameter

`docker run -e FOOBAR env_test`

3. See that the container never becomes healthy with `docker ps`

**Describe the results you received:**

The container never becomes healthy when viewing it with `docker ps` and the docker logs contain the following lines:
```
May 04 10:50:23 cake dockerd[7933]: time="2017-05-04T10:50:23.949217373-06:00" level=warning msg="Health check for container ce3525412a4a1b31cfb9de8ae8055cb1e9606635f24a6eb0726ec4ece03c3c90 error: rpc error: code = 2 desc = oci runtime error: exec failed: container_linux.go:247: starting container process caused \"invalid environment 'FOOBAR'\"\n"
```

**Describe the results you expected:**
The healthcheck would pass and the container would become healthy.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 548
Server Version: 17.04.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1354
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: ckkq4hik2h150qnxm238w7ik6
 Is Manager: true
 ClusterID: 17gwd1qpd0rtyvdzwbvi810i7
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.19.9.98
 Manager Addresses:
  172.19.9.98:2377
Runtimes: runc
Default Runtime: runc
Init Binary: 
containerd version: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-75-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 22.51GiB
Name: cake
ID: A4YB:3DTQ:L64X:TUHB:FZKK:NYL6:YW62:HCEX:SH3S:VKWM:B65H:ZTQ7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Ubuntu 16.04

### PR DESCRIPTION
**Pull request by boaz0**
fixes/closes #33021 

**- What I did**

Add environment variables and set the environment variables for the health check.

**- How I did it**

Use `ReplaceOrAppendEnvValues`, `CreateDaemonEnvironment` and `setupLinkedContainers` to collect all of them and to set them correctly.

**- How to verify it**

Run the following test:

```bash
$ TESTFLAGS='-check.f TestUnsetEnvVarHealthCheck' hack/make.sh test-integration-cli
```

### CODE REVIEW

File: opts/env_test.go [hunk 1]
```diff
@@ -4,27 +4,38 @@ import (
...
-		"PATH":                fmt.Sprintf("PATH=%v", os.Getenv("PATH")),
-		"PATH=something":      "PATH=something",
-		"asd!qwe":             "asd!qwe",
-		"1asd":                "1asd",
-		"123":                 "123",
-		"some space":          "some space",
-		"  some space before": "  some space before",
-		"some space after  ":  "some space after  ",
+		"a":                  "a=",
+		"asd!qwe":            "asd!qwe=",
```
Comment (mlaventure) [inline opts/env_test.go]: This and the one below are not valid environment variable name usually

Comment (ripcurld0) [inline opts/env_test.go]: Agree but when doing `os.Setenv("asd!qwe", "myvalue")` it worked.
So I am not sure what to do here. In addition, I have no idea whether that name is valid on windows too?

**Comment by ripcurld0** [comment]
OK, now I realized that if the variable isn't set, it should be discarded (DockerSuite.TestRunEnvironmentErase).

I am going to change this now.

File: opts/env_test.go (hunk 1 above)
Comment (mlaventure) [inline opts/env_test.go]: I think technically any character except `=` and `NUL` are allowed in the variable names, but implementations usually restrict themselves to the set of Portable characters as defined by POSIX -> `[a-zA-Z_][a-zA-Z0-9_]*`

Comment (thaJeztah) [inline opts/env_test.go]: Nope; we don't validate env var names; IIRC there's no official standard for it, and we did validation in the past, but broke people that used systems that don't comply with those rules.

We leave the validation to the container itself to tell if it's valid or not

Comment (mlaventure) [inline opts/env_test.go]: Fair enough, I have yet to use a shell that accept non POSIX names though :)

Comment (thaJeztah) [inline opts/env_test.go]: for reference, see https://github.com/moby/moby/pull/16608

Comment (mlaventure) [inline opts/env_test.go]: I was about to quote the open group page this :+1:

**Review by mlaventure — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Failure in `docker-py` looks related;

```
19:09:27 =================================== FAILURES ===================================
19:09:27 ______ CreateContainerTest.test_create_with_environment_variable_no_value ______
19:09:27 /docker-py/tests/integration/api_container_test.py:370: in test_create_with_environment_variable_no_value
19:09:27     environment={'Foo': None, 'Other': 'one', 'Blank': ''},
19:09:27 /docker-py/docker/api/container.py:446: in create_container
19:09:27     return self.create_container_from_config(config, name)
19:09:27 /docker-py/docker/api/container.py:457: in create_container_from_config
19:09:27     return self._result(res, True)
19:09:27 /docker-py/docker/api/client.py:220: in _result
19:09:27     self._raise_for_status(response)
19:09:27 /docker-py/docker/api/client.py:216: in _raise_for_status
19:09:27     raise create_api_error_from_http_exception(e)
19:09:27 /docker-py/docker/errors.py:30: in create_api_error_from_http_exception
19:09:27     raise cls(e, response=response, explanation=explanation)
19:09:27 E   APIError: 500 Server Error: Internal Server Error ("environment variable does not exists")
19:09:27 ======== 1 failed, 205 passed, 31 skipped, 2 xfailed in 244.48 seconds =========
19:09:27 ---> Making bundle: .integration-daemon-stop (in bundles/17.06.0-dev/test-docker-py)
19:09:27 +++++ cat bundles/17.06.0-dev/test-docker-py/docker.pid
19:09:27 ++++ kill 31590
19:09:28 ++++ /etc/init.d/apparmor stop
```

**Comment by ripcurld0** [comment]
@thaJeztah yes it is. I am working on fixing it 😺

**Comment by thaJeztah** [comment]
Did some testing with this. First ran a regular container, no healthcheck, and variations of ENV vars, to check the behavior;

```bash
$ docker build -t repro-33249 -<<EOF
FROM busybox
ENV IWILLBEUNSET=thiswasmyvalue
EOF


$ FROMENV=hello docker run -dit --name foobar \
  -e IDONOTEXIST \
  -e IWILLBEUNSET \
  -e IAMEMPTY= \
  -e FROMENV \
  repro-33249

$ docker inspect --format '{{json .Config.Env }}' foobar | jq .
[
  "IDONOTEXIST",
  "IWILLBEUNSET",
  "IAMEMPTY=",
  "FROMENV=hello",
  "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
]

$ docker exec foobar env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=211bf084aab2
IAMEMPTY=
FROMENV=hello
HOME=/root
```

So; `-e SOMEVAR` will be ;

- `empty` if followed by `=`
- get its value from the current environment (if exists)
- `unset` if it existed in the image, but does not exist in the environment

_Basically, in `.Config.Env`, `"SOMEVAR"` means: `unset` the env-var_


### With this patch (adding a healthcheck):

```bash
$ docker build -t repro-33249 -<<EOF
FROM busybox
ENV IWILLBEUNSET=thiswasmyvalue
HEALTHCHECK --interval=1s --timeout=5s --retries=5 CMD env; exit 0;
EOF

$ FROMENV=hello docker run -dit --name foobar \
  -e IDONOTEXIST \
  -e IWILLBEUNSET \
  -e IAMEMPTY= \
  -e FROMENV \
  repro-33249

$ docker inspect --format '{{json .Config.Env }}' foobar | jq .
[
  "IDONOTEXIST",
  "IWILLBEUNSET",
  "IAMEMPTY=",
  "FROMENV=hello",
  "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
]

$ docker exec foobar env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=2c0f5e0add62
IAMEMPTY=
FROMENV=hello
HOME=/root
```

All of the above are the same (as expected).


Checking the `healthcheck`;

```bash
$ docker inspect --format '{{json .State.Health.Log }}' foobar | jq .
[
  {
    "Start": "2017-05-19T11:27:41.45071088Z",
    "End": "2017-05-19T11:27:41.505086124Z",
    "ExitCode": 0,
    "Output": "SHLVL=1\nHOME=/root\nPATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\nFROMENV=hello\nIAMEMPTY=\nPWD=/\n"
  }
]
```

(expanding, and sorting a bit)

```
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
IAMEMPTY=
FROMENV=hello
HOME=/root
SHLVL=1
PWD=/
```

**Interesting; `HOSTNAME` is present in `docker exec`, but not present in the healthcheck's environment**

**Comment by thaJeztah** [comment]
So, as far as I can see, this works as expected.

I wonder if we (separate from this PR) should look into _why_ `HOSTNAME` is not present in the health check's environment.

**Review by thaJeztah — APPROVED** [review]
LGTM, but @mlaventure PTAL if this is all good. Also wondering how a regular `run` / `exec` differs from this health check

**Comment by ripcurld0** [comment]
@thaJeztah it is because `HOSTNAME` and `PATH` are created if `tty` is `TERM`:
https://github.com/moby/moby/blob/master/container/container_unix.go#L52-L55

**Comment by ripcurld0** [comment]
@thaJeztah I have an idea what happened! Thanks for noticing the "HOSTNAME" thing.

**Review by mlaventure — APPROVED** [review]
LGTM.

Thanks for the fix!

**Pull request merged**